torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,534 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a DAG that dynamically maintains a topological order as
+// edges/vertices are added and removed.
+//
+// [Example applications]
+// - Let's say you have a DAG where each vertex is black or red. How do we
+//   merge black nodes that are directly connected by contracting the
+//   edge between them while still maintaining the DAG and a topological order?
+//   Use contractEdge().
+// - Let's say you have a DAG where each vertex is a Node* and the edges represent
+//   data dependencies. We wish to determine if adding a new Node* with certain
+//   data dependencies (or moving an existing one to use new dependencies) is valid.
+//   Use DynamicDAG::addEdge() to add the new data dependencies to the DAG:
+//   it will either find a valid reordering of the DAG's topological order or throw
+//   if the resulting DAG is invalid.
+//
+// The implementation is based off of the PK algorithm in the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+// It is summarized in [Edge addition] (see DynamicDAG<T>::addEdge)
+
+template <typename T> struct Vertex;
+template <typename T> struct DynamicDAG;
+template <typename T> using vertex_list = std::vector<Vertex<T>*>;
+template <typename T> using unique_vertex = std::unique_ptr<Vertex<T>>;
+
+enum DFSDirection {forward, backward};
+
+// Because our graphs shouldn't fan out or in very much,
+// we use std::vector<Vertex<T>*> to record edges.
+// In all of the complexity analysis it is assumed that
+// inserting, erasing, and finding take constant time.
+template <typename T>
+struct EdgeData {
+  // These are both sorted in topological order because iterating
+  // in topological order is useful.
+  vertex_list<T> in_edges;
+  vertex_list<T> out_edges;
+
+  static void insert(vertex_list<T>& lst, Vertex<T>* v) {
+    // Keep the list sorted.
+    // We can do the same thing (binary search) for erase() and has()
+    // but I'm not sure if it will perform better especially since
+    // our lists should be small.
+    lst.insert(std::upper_bound(lst.begin(), lst.end(), v,
+          [](Vertex<T>* x, Vertex<T>* y) { return x->ord < y->ord; }), v);
+  }
+
+  static void erase(vertex_list<T>& lst, Vertex<T>* v) {
+    lst.erase(std::find(lst.begin(), lst.end(), v));
+  }
+
+  static bool has(vertex_list<T>& lst, Vertex<T>* v) {
+    return std::find(lst.begin(), lst.end(), v) != lst.end();
+  }
```

Comment:
Sounds good!

Reply:
So actually my current implementation of the sorted_vertex_list is incorrect because `reorder(...)` can shift vertices around such that adjacency lists are no longer in order...

Going to think about this for a while

Reply:
Resolved, I wrote a wrapper struct